src/bun.js/bindings/ZigGlobalObject.cpp

```diff
@@ -1779,6 +1784,75 @@ JSC_DEFINE_HOST_FUNCTION(jsReceiveMessageOnPort, (JSGlobalObject * lexicalGlobal
     return JSC::JSValue::encode(jsUndefined());
 }
 
+JSC_DEFINE_HOST_FUNCTION(functionDomainToASCII, (JSC::JSGlobalObject * globalObject, JSC::CallFrame* callFrame))
+{
+    auto& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    auto arg0 = callFrame->argument(0);
+    if (!arg0.isString()) {
+        throwTypeError(globalObject, scope, "the \"domain\" argument is missing or not a string"_s);
+        return JSC::JSValue::encode(jsUndefined());
+    }
+
+    auto domain = arg0.toWTFString(globalObject);
+    if (domain.isNull())
+        return JSC::JSValue::encode(jsUndefined());
+    if (domain.is8Bit())
+        return JSC::JSValue::encode(arg0);
+
+    constexpr static int allowedNameToASCIIErrors = UIDNA_ERROR_EMPTY_LABEL | UIDNA_ERROR_LABEL_TOO_LONG | UIDNA_ERROR_DOMAIN_NAME_TOO_LONG | UIDNA_ERROR_LEADING_HYPHEN | UIDNA_ERROR_TRAILING_HYPHEN | UIDNA_ERROR_HYPHEN_3_4;
+    constexpr static size_t hostnameBufferLength = 2048;
+
+    auto encoder = &WTF::URLParser::internationalDomainNameTranscoder();
+    UChar hostnameBuffer[hostnameBufferLength];
+    UErrorCode error = U_ZERO_ERROR;
+    UIDNAInfo processingDetails = UIDNA_INFO_INITIALIZER;
+    int32_t numCharactersConverted = uidna_nameToASCII(encoder, StringView(domain).characters16(), domain.length(), hostnameBuffer, hostnameBufferLength, &processingDetails, &error);
+
+    if (U_SUCCESS(error) && !(processingDetails.errors & ~allowedNameToASCIIErrors) && numCharactersConverted) {
+        return JSC::JSValue::encode(JSC::jsString(vm, WTF::String(hostnameBuffer, numCharactersConverted)));
+    }
+    throwTypeError(globalObject, scope, "domainToASCII failed"_s);
+    return JSC::JSValue::encode(jsUndefined());
```

Comment:
Instead of throwing Node returns an empty string:
https://github.com/nodejs/node/blob/d96606d9aab35baf9e5c77cf9312145c3ae6086b/src/node_url.cc#L76-L97
For example `url.domainToUnicode('@')` returns `''`.

Reply:
I'd seen that code but wasn't totally sure of its error conditions. I'll try the `'@'` in my version and compare

Reply:
rn it doesnt throw, it returns back `'@'`

Reply:
found the list https://url.spec.whatwg.org/#forbidden-host-code-point